spdx/identifier.go

```diff
@@ -66,37 +125,57 @@ func (d *DocElementID) UnmarshalJSON(data []byte) error {
 	}
 
 	// handle SPDXRef-
-	idFields = strings.SplitN(idStr, "SPDXRef-", 2)
-	if len(idFields) != 2 {
-		return fmt.Errorf("failed to parse SPDX Identifier '%s'", idStr)
+	err := d.ElementRefID.FromString(idStr)
+	if err != nil {
+		return err
 	}
 
-	d.ElementRefID = ElementID(idFields[1])
-
 	return nil
 }
 
-// MarshalJSON converts the receiver into a slice of bytes representing a DocElementID in string form.
-// This function is also used when marshalling to YAML
-func (d DocElementID) MarshalJSON() ([]byte, error) {
+// MarshalString converts the receiver into a string representing a DocElementID.
+// This is used when writing a spreadsheet SPDX file, for example.
+func (d DocElementID) String() string {
 	if d.DocumentRefID != "" && d.ElementRefID != "" {
-		return json.Marshal(fmt.Sprintf("DocumentRef-%s:SPDXRef-%s", d.DocumentRefID, d.ElementRefID))
+		return fmt.Sprintf("DocumentRef-%s:%s", d.DocumentRefID, d.ElementRefID)
+	} else if d.DocumentRefID != "" {
+		return fmt.Sprintf("DocumentRef-%s", d.DocumentRefID)
 	} else if d.ElementRefID != "" {
-		return json.Marshal(fmt.Sprintf("SPDXRef-%s", d.ElementRefID))
+		return d.ElementRefID.String()
 	} else if d.SpecialID != "" {
-		return json.Marshal(d.SpecialID)
+		return d.SpecialID
 	}
 
-	return []byte{}, fmt.Errorf("failed to marshal empty DocElementID")
+	return ""
 }
 
-// TODO: add equivalents for LicenseRef- identifiers
+// UnmarshalJSON takes a SPDX Identifier string parses it into a DocElementID struct.
+// This function is also used when unmarshalling YAML
+func (d *DocElementID) UnmarshalJSON(data []byte) error {
+	// SPDX identifier will simply be a string
+	idStr := string(data)
+	idStr = strings.Trim(idStr, "\"")
 
-// MakeDocElementID takes strings (without prefixes) for the DocumentRef-
-// and SPDXRef- identifiers, and returns a DocElementID. An empty string
-// should be used for the DocumentRef- portion if it is referring to the
-// present document.
+	return d.FromString(idStr)
+}
+
+// MarshalJSON converts the receiver into a slice of bytes representing a DocElementID in string form.
+// This function is also used when marshalling to YAML
+func (d DocElementID) MarshalJSON() ([]byte, error) {
+	if err := d.Validate(); err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(d.String())
+}
+
+// MakeDocElementID takes strings for the DocumentRef- and SPDXRef- identifiers (these prefixes will be stripped if present),
+// and returns a DocElementID.
+// An empty string should be used for the DocumentRef- portion if it is referring to the present document.
 func MakeDocElementID(docRef string, eltRef string) DocElementID {
+	docRef = strings.Replace(docRef, "DocumentRef-", "", 1)
+	eltRef = strings.Replace(eltRef, "SPDXRef-", "", 1)
```

Comment:
I'm a bit uncomfortable with having this function be able to take strings which _might or might not_ have the `DocumentRef-` or `SPDXRef-` prefixes.

This is an edge case probably, but in an SPDX Document you could have two different SPDX elements with the following SPDX Identifiers:
* `SPDXRef-File1`
* `SPDXRef-SPDXRef-File1`

I wouldn't recommend anyone actually build an SPDX Document this way, but I think it would be syntactically valid.

So if we've got functions that are optionally stripping the first instance of `SPDXRef-` that it sees, I feel like that will lead to unexpected behavior for edge cases like the one above. I'm having trouble putting my finger on a concrete example, but I guess I'm thinking that it'll be easier to avoid doing the wrong thing if a function like this always requires its inputs to either have the prefix or omit the prefix.

Reply:
I think your example should just work, due to the `, 1)` at the end of the `strings.Replace(...)` call.

```
strings.Replace("SPDXRef-File1", "SPDXRef-", "", 1) -> "File1"
strings.Replace("SPDXRef-SPDXRef-File1", "SPDXRef-", "", 1) -> "SPDXRef-File1"
strings.Replace("SPDXRef-DocumentRef-File1", "SPDXRef-", "", 1) -> "DocumentRef-File1"
strings.Replace("File1-SPDXRef", "SPDXRef-", "", 1) -> "File1-SPDXRef"
```

If we can come up with edge cases we're worried about, I can add test cases for them to ensure that this function is handling them the way we'd want.

But as I mentioned in another comment, this is a situation we have run into repeatedly with real SPDX documents from users, where sometimes they just omit the `SPDXRef-` prefix on their identifiers, and it's not sustainable for us to fix that manually every time it happens. I think this is a case where the user clearly meant to add the prefix but just forgot/missed it and it's safe for us to make an assumption.